File: data-structures/c-or-cpp/SinglyLinkedList.cpp
#include <iostream>

template <typename T>
class Node {
    public:
        T data;
        Node *next;

        Node(T data)
            : data(data) , next(nullptr) {};
};

template <typename T>
class SinglyLinkedList {
    private:
        Node<T>* head;

    public:
        SinglyLinkedList()
            : head(nullptr){}

        void push(T data) {
            Node<T> *temp = new Node<T>(data);
            if(this->head == nullptr) {
                this->head = temp;
                return;
            }

            temp->next = this->head;
            this->head = temp;
        }

        T pop() {
            if(this->head == nullptr) {
                std::cout<<"List is Empty"<<std::endl;
                exit(1);
            }
            T data = this->head->data;
            this->head = this->head->next;
            return data;
        }

        void printList() const {
            if(this->head == nullptr) {
                std::cout<<"List is Empty"<<std::endl;
            } else {
                Node<T> *itr = this->head;
                while(itr != nullptr) {
                    std::cout<<itr->data<<" ";
                    itr = itr->next;
                }
                std::cout<<std::endl;
            }
        }
};

int main() {
    SinglyLinkedList<int> l;
    l.push(1);
    l.push(2);
    l.push(3);
    l.pop();
    l.printList();
    return 0;
}

File: linked-lists/java/SinglyLinkedLIst.java
class Node<T> {
    T data;
    Node<T> next;

    Node(T data) {
        this.data = data;
        this.next = null;
    }
}

public class SinglyLinkedList<T> {
    Node<T> head;

    SinglyLinkedList() {
        this.head = null;
    }

    void insertAtHead(T data) {
        Node<T> newNode = new Node<T>(data);
        if(this.head == null) {
            this.head = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
    }

    T removeAtHead() {
        if(this.head == null) {
            System.out.println("Underflow");
            System.exit(1);
        }
        T data = this.head.data;
        this.head = this.head.next;
        return data;
    }

    void printList() {
        if(this.head == null) {
            System.out.println("List is Empty");
        } else {
            Node<T> itr = this.head;
            while(itr != null) {
                System.out.print(itr.data + " ");
                itr = itr.next;
            }
            System.out.println("");
        }
    }

    // Test Code
    public static void main(String... args) {
        SinglyLinkedList<String> l = new SinglyLinkedList<String>();
        l.insertAtHead("abc");
        l.insertAtHead("pqr");
        l.insertAtHead("xyz");
        l.removeAtHead();
        l.printList();
    }
}
